batch/batch/front_end/front_end.py

```diff
@@ -338,52 +338,67 @@ async def _get_job_log_from_record(app, batch_id, job_id, record):
                 return None
             raise
 
-    if state in ('Error', 'Failed', 'Success'):
-        file_store: FileStore = app['file_store']
-        batch_format_version = BatchFormatVersion(record['format_version'])
+    if state in ('Pending', 'Ready', 'Creating'):
+        return None
 
-        async def _read_log_from_gcs(task):
-            try:
-                data = await file_store.read_log_file(
-                    batch_format_version, batch_id, job_id, record['attempt_id'], task
-                )
-            except FileNotFoundError:
-                id = (batch_id, job_id)
-                log.exception(f'missing log file for {id} and task {task}')
-                data = 'ERROR: could not find log file'
-            return task, data
+    if state == 'Cancelled' and record['last_cancelled_attempt_id'] is None:
+        return None
 
-        spec = json.loads(record['spec'])
-        tasks = []
+    assert state in ('Error', 'Failed', 'Success', 'Cancelled')
 
-        has_input_files = batch_format_version.get_spec_has_input_files(spec)
-        if has_input_files:
-            tasks.append('input')
+    batch_format_version = BatchFormatVersion(record['format_version'])
+    attempt_id = record['attempt_id'] or record['last_cancelled_attempt_id']
+    assert attempt_id
 
-        tasks.append('main')
+    file_store: FileStore = app['file_store']
+    batch_format_version = BatchFormatVersion(record['format_version'])
 
-        has_output_files = batch_format_version.get_spec_has_output_files(spec)
-        if has_output_files:
-            tasks.append('output')
+    async def _read_log_from_gcs(task):
+        try:
+            data = await file_store.read_log_file(batch_format_version, batch_id, job_id, attempt_id, task)
+        except FileNotFoundError:
+            id = (batch_id, job_id)
+            log.exception(f'missing log file for {id} and task {task}')
+            data = 'ERROR: could not find log file'
+        return task, data
+
+    spec = json.loads(record['spec'])
+    tasks = []
+
+    has_input_files = batch_format_version.get_spec_has_input_files(spec)
+    if has_input_files:
+        tasks.append('input')
 
-        return dict(await asyncio.gather(*[_read_log_from_gcs(task) for task in tasks]))
+    tasks.append('main')
 
-    return None
+    has_output_files = batch_format_version.get_spec_has_output_files(spec)
+    if has_output_files:
+        tasks.append('output')
+
+    return dict(await asyncio.gather(*[_read_log_from_gcs(task) for task in tasks]))
 
 
 async def _get_job_log(app, batch_id, job_id):
     db: Database = app['db']
 
     record = await db.select_and_fetchone(
         '''
```

Comment:
Can you give me some context on why the DB needs to change? It seems to me that the issue is that JobDeletedError was skipping the cleanup in the worker. Is there more to the bug?

Reply:
The other issue was we weren't uploading status, logs for cancelled jobs on the worker.